design/batch_retrieval_design.md

```diff
@@ -2,19 +2,140 @@
+The following table outlines possible solutions to overcome this:
 
-|      | Solution                                                     | Pros                                                         | Cons                                                         | Effort (T-shirt size) |
-| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------- |
-| 1    | *Client side:* <br />For each K8s Secret, perform a new Batch request | - No server side changes / no breaking changes               | Load on server with extra calls (*)                          | S, 5 days             |
-| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Better / straightforward design for how batch endpoints | - Requires both client/server implementation<br />- Need to handle backwards compatibility | M, 10 days            |
-| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                                       | -                     |
+|      | Solution                                                     | Pros                                                         | Cons                                               | Effort (T-shirt size) |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | -------------------------------------------------- | --------------------- |
+| 1    | *Client side:* <br />For each K8s Secret, perform a new Batch request | - No server side changes / no breaking changes               | Load on server with extra calls (*)                | S, 5 days             |
+| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Best practice / straightforward design for batch endpoint | - Requires both client/server implementation<br /> | M, 10 days            |
+| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                             | -                     |
 
 (*) Fallback solution: use this solution as a safety net, only if the original Batch Retrieval request fails
 
 (**) This solution can be broken up in two: 1. Create a new API endpoint or 2. use existing one (return 409 response)
 
+*Decision:* We decided to go with #2 and create our own API. The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
```

Comment:
i think that this design is lacking info for how the new endpoint will be implemented. did you consider using the current batch retrieval api? if so, why did you decide not to use it?

Reply:
> i think that this design is lacking info for how the new endpoint will be implemented.

As in adding it to `routes.rb` etc? Based on other docs I have seen on confluence, I actually think that is a bit too low level. Do you think differently?

> did you consider using the current batch retrieval api? if so, why did you decide not to use it?

Indeed! We chose to add another Batch Retrieval API for the following reasons:
1. Add onto the existing batch endpoint would lead to changing behaviors for our needs which would have cascading impacts on other teams which is not what we wanted to accomplish. Also they are two separate use-cases 1️⃣  partial success or 2️⃣ fail fast
2. In terms of performance, having a long string of variableIds as query params impacts performance verses sending those variableIds in the body of the request Improves.  From what I read, query params are used for filtering (search/pagination or `?withComments`).  Since we are anticipating heavy parameter usage, we decided to send all this data in the body of a POST request. 

This is probably a question a lot of people will ask so I will add this to the doc

Reply:
i still think that we can use the existing API. there are 2 differences between the two APIs:
1. the required variables are passed in the body instead of in a query param - this can be handled by reading the body _and_ reading the query param. we can decide if we want to merge the lists or reject a request that has both inputs but we can still use the existing one with backwards compatibility.
2. The response body is quite the same iirc, where the only difference is that we don't return only the value but also a `status` field. So the object is still a key-value json where the key is the variable-id. This response code will also change accordingly as the new API will not fail if only part of the variables will fail. This can be filtered by a query_param (e.g `partial`) that if present we will return the new API response. This is also backwards-compatible.

Reply:
@abrahamko ^

Reply:
Why do we need the statuses for each variable?
Few questions regarding this:
1. I believe that 403 status is problematic from security aspect, no? I have permissions we need to 
2. Generally, is secret provider will act differently for specific statues? or is it only for supportability?

Reply:
@InbalZilberman 
1️⃣  From our docs, I see that Batch can also [returns 403](https://docs.conjur.org/Latest/en/Content/Developer/Conjur_API_Batch_Retrieve.htm). I am not saying this is right behavior because we did it before. But it is important to note that only the nested, sub-requests return the error codes **NOT** the batch request itself. It is possible that each secret returns 403 error but the batch request itself will return 200 status OK. IMO since we are just an API endpoint, it is up to the צורך to decide if he wants to cascade that error to his application. @sashaCher what say you?

2️⃣ To know if we have failed or succeeded and on which exact secret. I think it gives nice insight into the status of each secret that was retrieved from Conjur. By adding `status` in the response, it makes it easy for the Secrets Provider to just extract the error code and percolate up that error, manipulating the result how we see fit (returning a log "Updating x out of y secrets", etc)

Reply:
I agree that @sashaCher should provide a guidance for it

Reply:
`403` is acceptable error code **if and only if** the role has `READ` permission on given variable but has no `EXECUTE` one.
`404` is the right error code for all other use cases.

Reply:
How individual status codes are translated to the API status code is a functional decision.
From my perspective it should be 200 as far as there's a secret in response that has 200...